src/py21cmmc/likelihood.py

```diff
@@ -965,3 +965,139 @@ def define_noise(self, ctx, model):
+                    # it's larger than the frequency of the minimum, so it's the upper bound of the fwhm
+                    # then use the boundary to be the lower one
+                    freq_r = freqs_hm[0]
+                    freq_l = frequencies[0]
+                else:
+                    # it's smaller than the frequency of the minimum, so it's the lower bound of the fwhm
+                    # then use the boundary to be the upper one
+                    freq_l = freqs_hm[0]
+                    freq_r = frequencies[-1]
+            elif len(freqs_hm) > 2:
+                # therea are more two of them, need to find the closest two
+                freq_1 = freqs_hm[np.argmin(np.abs(freqs_hm - freq_tb_min))]
+                if freq_1 < freq_tb_min:
+                    # the closest one is smaller than the frequency of the minimum, so it's the lower bound of the fwhm
+                    freq_l = freq_1
+                    freq_rs = freqs_hm[freqs_hm > freq_tb_min]
+                    # find the rest which are larger than the frequency of the minimum
+                    if len(freq_rs) > 0:
+                        # the smallest should be the upper bound of the fwhm
+                        freq_r = freq_rs[0]
+                    else:
+                        # if none, use the boundary
+                        freq_r = frequencies[-1]
+                else:
+                    # the closest one is larger than the frequency of the minimum, so it's the upper bound of the fwhm
+                    freq_r = freq_1
+                    freq_ls = freqs_hm[freqs_hm < freq_tb_min]
+                    # find the rest which are smaller than the frequency of the minimum
+                    if len(freq_ls) > 0:
+                        # the largest should be the lower bound of the fwhm
+                        freq_l = freq_ls[-1]
+                    else:
+                        # if none, use the boundary
+                        freq_l = frequencies[0]
+            if len(freqs_hm) == 0:
+                results["fwhm"] = None
+            else:
+                results["fwhm"] = freq_r - freq_l
+        return results
+
+    def computeLikelihood(self, model):
+        """
+        Compute the likelihood, given the lightcone output from 21cmFAST.
+
+        Parameters
+        ----------
+
+        model : list of dicts
+            Exactly the output of :meth:`simulate`.
+
+        Returns
+        -------
+
+        lnl : float
+            The log-likelihood for the given model.
+
+        """
+
+        if model["freq_tb_min"] is None:
+            return -np.inf
+
+        if self.use_width:
+            if model["fwhm"] is None:
+                return -np.inf
+            else:
+                return -0.5 * np.square(
+                    (model["freq_tb_min"] - self.freq_edges) / self.freq_err_edges
+                ) + -0.5 * np.square(model["fwhm"] - self.fwhm_edges) / (
```

Comment:
You need to be careful when using this approximation for asymmetric errors. You need to put in checks that the denominator doesn't go negative